File: src/ios/TwilioVideoViewController.m
//
//  TwilioVideoViewController.m
//
//  Copyright © 2016-2017 Twilio, Inc. All rights reserved.
//

@import TwilioVideo;
#import "TwilioVideoViewController.h"


#import <Foundation/Foundation.h>

@interface PlatformUtils : NSObject

+ (BOOL)isSimulator;

@end

@implementation PlatformUtils

+ (BOOL)isSimulator {
#if TARGET_IPHONE_SIMULATOR
    return YES;
#endif
    return NO;
}

@end

@interface TwilioVideoViewController () <UITextFieldDelegate, TVIParticipantDelegate, TVIRoomDelegate, TVIVideoViewDelegate, TVICameraCapturerDelegate>

#pragma mark Video SDK components

@property (nonatomic, strong) TVIParticipant *viewedParticipant;
@property (nonatomic, weak) TVIVideoView *remoteView;
@property (nonatomic, strong) TVIRoom *room;

#pragma mark UI Element Outlets and handles



@property (nonatomic, weak) IBOutlet UIButton *disconnectButton;
@property (nonatomic, weak) IBOutlet UILabel *messageLabel;

// CS-69: Weak timer or else we get a crash on invalidate
@property (nonatomic, weak) NSTimer *timer;

@end

@implementation TwilioVideoViewController
const double ANIMATION_DURATION = 0.4;
const double TIMER_INTERVAL = 4;
const UIInterfaceOrientation defaultOrientation = UIInterfaceOrientationLandscapeRight;
@synthesize delegate;

#pragma mark - UIViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self logMessage:[NSString stringWithFormat:@"TwilioVideo v%@", [TwilioVideo version]]];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

    // Lock to landscape orientation.
    UIInterfaceOrientation orientation = UIDevice.currentDevice.orientation;
    if (!UIInterfaceOrientationIsLandscape(orientation)) {
        orientation = defaultOrientation;
    }
    [[UIDevice currentDevice] setValue: [NSNumber numberWithInteger: orientation] forKey:@"orientation"];

    // Start icon timer
    [self startTimer];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
	
    // Stop any timer
    [self stopTimer];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if(self.disconnectButton.hidden) {
        self.disconnectButton.hidden = !self.disconnectButton.hidden;
        [self showDisconnectButton];
        [self startTimer];
    } else {
        [self resetTimer];
    }
}

-(void)hideDisconnectButton {
    [UIView animateWithDuration:ANIMATION_DURATION delay:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
        self.disconnectButton.layer.opacity = 0.0f;
    } completion: ^(BOOL finished) {
        if(finished) {
            self.disconnectButton.hidden = !self.disconnectButton.hidden;
        }
    }];
}

-(void)showDisconnectButton {
    [UIView animateWithDuration:ANIMATION_DURATION delay:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
        self.disconnectButton.layer.opacity = 1.0f;
    } completion: nil];
}

-(void)startTimer {
    self.timer = [NSTimer scheduledTimerWithTimeInterval:TIMER_INTERVAL target:self selector:@selector(hideDisconnectButton) userInfo:nil repeats:NO];
}

-(void)resetTimer {
    [self stopTimer];
    [self startTimer];
}

-(void)stopTimer {
    [self.timer invalidate];
    self.timer = nil;
}  

#pragma mark - Public
    
- (void)connectToRoom:(NSString*)room {
    [self showRoomUI:YES];
    
    if ([self.accessToken isEqualToString:@"TWILIO_ACCESS_TOKEN"]) {
        [self logMessage:[NSString stringWithFormat:@"Fetching an access token"]];
        [self showRoomUI:NO];
    } else {
        [self doConnect:room];
    }
}

- (IBAction)disconnectButtonPressed:(id)sender {
    [self stopTimer];
    [self.room disconnect];
    [self.delegate dismiss];
}

#pragma mark - Private

- (void)doConnect:(NSString*)room {
    if ([self.accessToken isEqualToString:@"TWILIO_ACCESS_TOKEN"]) {
        //   [self logMessage:@"Please provide a valid token to connect to a room"];
        return;
    }

    TVIConnectOptions *connectOptions = [TVIConnectOptions optionsWithToken:self.accessToken
                                                                      block:^(TVIConnectOptionsBuilder * _Nonnull builder) {
                                                                          
                                                                          // The name of the Room where the Client will attempt to connect to. Please note that if you pass an empty
                                                                          // Room `name`, the Client will create one for you. You can get the name or sid from any connected Room.
                                                                          builder.roomName = room;
                                                                      }];
    
    // Connect to the Room using the options we provided.
    self.room = [TwilioVideo connectWithOptions:connectOptions delegate:self];
    
    //   [self logMessage:[NSString stringWithFormat:@"Attempting to connect to room %@", room]];
}

- (void)setupRemoteView {
    // Creating `TVIVideoView` programmatically
    TVIVideoView *remoteView = [[TVIVideoView alloc] init];
    
    // `TVIVideoView` supports UIViewContentModeScaleToFill, UIViewContentModeScaleAspectFill and UIViewContentModeScaleAspectFit
    // UIViewContentModeScaleAspectFit is the default mode when you create `TVIVideoView` programmatically.
    self.remoteView.contentMode = UIViewContentModeScaleAspectFit;
    
    [self.view insertSubview:remoteView atIndex:0];
    self.remoteView = remoteView;
    
    NSLayoutConstraint *centerX = [NSLayoutConstraint constraintWithItem:self.remoteView
                                                               attribute:NSLayoutAttributeCenterX
                                                               relatedBy:NSLayoutRelationEqual
                                                                  toItem:self.view
                                                               attribute:NSLayoutAttributeCenterX
                                                              multiplier:1
                                                                constant:0];
    [self.view addConstraint:centerX];
    NSLayoutConstraint *centerY = [NSLayoutConstraint constraintWithItem:self.remoteView
                                                               attribute:NSLayoutAttributeCenterY
                                                               relatedBy:NSLayoutRelationEqual
                                                                  toItem:self.view
                                                               attribute:NSLayoutAttributeCenterY
                                                              multiplier:1
                                                                constant:0];
    [self.view addConstraint:centerY];
    NSLayoutConstraint *width = [NSLayoutConstraint constraintWithItem:self.remoteView
                                                             attribute:NSLayoutAttributeWidth
                                                             relatedBy:NSLayoutRelationEqual
                                                                toItem:self.view
                                                             attribute:NSLayoutAttributeWidth
                                                            multiplier:1
                                                              constant:0];
    [self.view addConstraint:width];
    NSLayoutConstraint *height = [NSLayoutConstraint constraintWithItem:self.remoteView
                                                              attribute:NSLayoutAttributeHeight
                                                              relatedBy:NSLayoutRelationEqual
                                                                 toItem:self.view
                                                              attribute:NSLayoutAttributeHeight
                                                             multiplier:1
                                                               constant:0];
    [self.view addConstraint:height];
}

// Reset the client ui status
- (void)showRoomUI:(BOOL)inRoom {
    // self.micButton.hidden = !inRoom;
    // self.disconnectButton.hidden = !inRoom;
    [UIApplication sharedApplication].idleTimerDisabled = inRoom;
}

- (void)cleanupRemoteParticipant {
    if (self.viewedParticipant) {
        if ([self.viewedParticipant.videoTracks count] > 0) {
            [self.viewedParticipant.videoTracks[0] removeRenderer:self.remoteView];
            [self.remoteView removeFromSuperview];
        }
        self.viewedParticipant = nil;
    }
}

- (void)logMessage:(NSString *)msg {
    NSLog(@"%@", msg);
}

#pragma mark - UITextFieldDelegate

#pragma mark - TVIRoomDelegate

- (void)didConnectToRoom:(TVIRoom *)room {
    // At the moment, this example only supports rendering one Participant at a time.
    
    // [self logMessage:[NSString stringWithFormat:@"Connected to room %@ as %@", room.name, room.localParticipant.identity]];
    [self logMessage:@"Waiting on participant to join"];
    self.messageLabel.text = self.remoteParticipantName;
    self.viewedParticipant = nil;
    for (TVIParticipant* participant in room.participants) {
        participant.delegate = self;
    }
}

- (void)room:(TVIRoom *)room didDisconnectWithError:(nullable NSError *)error {
    // [self logMessage:[NSString stringWithFormat:@"Disconncted from room %@, error = %@", room.name, error]];
    
    [self cleanupRemoteParticipant];
    self.room = nil;
    
    [self showRoomUI:NO];
    [self.delegate dismiss];
}

- (void)room:(TVIRoom *)room didFailToConnectWithError:(nonnull NSError *)error{
    //  [self logMessage:[NSString stringWithFormat:@"Failed to connect to room, error = %@", error]];
    
    self.room = nil;
    
    [self showRoomUI:NO];
    [self.delegate dismiss];
}

- (void)room:(TVIRoom *)room participantDidConnect:(TVIParticipant *)participant {
    participant.delegate = self;
    //   [self logMessage:[NSString stringWithFormat:@"Room %@ participant %@ connected", room.name, participant.identity]];
}

- (void)room:(TVIRoom *)room participantDidDisconnect:(TVIParticipant *)participant {
    // [self logMessage:[NSString stringWithFormat:@"Room %@ participant %@ disconnected", room.name, participant.identity]];
    if (self.viewedParticipant == participant) {
        [self logMessage:@"Participant disconnected"];
        [self cleanupRemoteParticipant];
        [self.delegate dismiss];
    }
}

#pragma mark - TVIParticipantDelegate

- (void)participant:(TVIParticipant *)participant addedVideoTrack:(TVIVideoTrack *)videoTrack {
    //   [self logMessage:[NSString stringWithFormat:@"Participant %@ added video track.", participant.identity]];
  
    if (self.viewedParticipant != participant) {
        [self cleanupRemoteParticipant];
        self.viewedParticipant = participant;
        [self setupRemoteView];
        [videoTrack addRenderer:self.remoteView];
    }
}

- (void)participant:(TVIParticipant *)participant removedVideoTrack:(TVIVideoTrack *)videoTrack {
    //   [self logMessage:[NSString stringWithFormat:@"Participant %@ removed video track.", participant.identity]];
    
    if (self.viewedParticipant == participant) {
        [videoTrack removeRenderer:self.remoteView];
        [self.remoteView removeFromSuperview];
        [self cleanupRemoteParticipant];
        [self.delegate dismiss];
        // TODO: This will kick us out....some ideas:
        //  1. Search for another participant with a video track (requires saving all participants or tracking in addedVideoTrack)
    }
}

- (void)participant:(TVIParticipant *)participant addedAudioTrack:(TVIAudioTrack *)audioTrack {
    //  [self logMessage:[NSString stringWithFormat:@"Participant %@ added audio track.", participant.identity]];
}

- (void)participant:(TVIParticipant *)participant removedAudioTrack:(TVIAudioTrack *)audioTrack {
    //  [self logMessage:[NSString stringWithFormat:@"Participant %@ removed audio track.", participant.identity]];
}

- (void)participant:(TVIParticipant *)participant enabledTrack:(TVITrack *)track {
    NSString *type = @"";
    if ([track isKindOfClass:[TVIAudioTrack class]]) {
        type = @"audio";
    } else {
        type = @"video";
    }
    //  [self logMessage:[NSString stringWithFormat:@"Participant %@ enabled %@ track.", participant.identity, type]];
}

- (void)participant:(TVIParticipant *)participant disabledTrack:(TVITrack *)track {
    NSString *type = @"";
    if ([track isKindOfClass:[TVIAudioTrack class]]) {
        type = @"audio";
    } else {
        type = @"video";
    }
    //  [self logMessage:[NSString stringWithFormat:@"Participant %@ disabled %@ track.", participant.identity, type]];
}

#pragma mark - TVIVideoViewDelegate

- (void)videoView:(TVIVideoView *)view videoDimensionsDidChange:(CMVideoDimensions)dimensions {
    NSLog(@"Dimensions changed to: %d x %d", dimensions.width, dimensions.height);
    [self.view setNeedsLayout];
}

@end

File: src/ios/TwilioVideoViewController.swift
//
//  TwilioVideoViewController.swift
//
//  Copyright (C) 2019 by Cape Productions, Inc.
//
// Converted from TwilioVideoViewController.m

import Foundation
import TwilioVideo
import UIKit

protocol TwilioVideoViewControllerDelegate: NSObjectProtocol {
    func dismiss()
    func onConnected(_ participantId: String?, participantSid: String?)
    func onDisconnected(_ participantId: String?, participantSid: String?)
}

private let ANIMATION_DURATION: Double = 0.4
private let TIMER_INTERVAL: Double = 4
private let defaultOrientation: UIInterfaceOrientation = .landscapeRight

@objc(TwilioVideoViewController) class TwilioVideoViewController: UIViewController, UITextFieldDelegate, TVIRemoteParticipantDelegate, TVIRoomDelegate, TVIVideoViewDelegate {
    weak var delegate: TwilioVideoViewControllerDelegate?
    var accessToken = ""
    var remoteParticipantName = ""

    func connect(toRoom room: String?) {
        guard let room = room else {
            logMessage("Invalid room")
            return
        }
        showRoomUI(true)

        if (accessToken == "TWILIO_ACCESS_TOKEN") {
            logMessage("Fetching an access token")
            showRoomUI(false)
        } else {
            doConnect(room)
        }
    }

// MARK: Video SDK components
    private var viewedParticipant: TVIRemoteParticipant?
    private weak var remoteView: TVIVideoView?
    private var room: TVIRoom?
// MARK: UI Element Outlets and handles


    @IBOutlet private weak var disconnectButton: UIButton!
    @IBOutlet private weak var messageLabel: UILabel!
    // CS-69: Weak timer or else we get a crash on invalidate
    private weak var timer: Timer?

// MARK: - UIViewController
    override func viewDidLoad() {
        super.viewDidLoad()

        logMessage("TwilioVideo v\(TwilioVideo.version())")
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        // Lock to landscape orientation.
        var orientation = UIInterfaceOrientation(rawValue: UIDevice.current.orientation.rawValue)
        if orientation?.isLandscape == false {
            orientation = defaultOrientation
        }
        UIDevice.current.setValue(NSNumber(value: orientation?.rawValue ?? 0), forKey: "orientation")

        // Start icon timer
        startTimer()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        // Stop any timer
        stopTimer()
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        if disconnectButton.isHidden {
            disconnectButton.isHidden = !disconnectButton.isHidden
            showDisconnectButton()
            startTimer()
        } else {
            resetTimer()
        }
    }

    @objc func hideDisconnectButton() {
        UIView.animate(withDuration: TimeInterval(ANIMATION_DURATION), delay: 0, options: .allowUserInteraction, animations: {
            self.disconnectButton.layer.opacity = 0.0
        }) { finished in
            if finished {
                self.disconnectButton.isHidden = !self.disconnectButton.isHidden
            }
        }
    }

    func showDisconnectButton() {
        UIView.animate(withDuration: TimeInterval(ANIMATION_DURATION), delay: 0, options: .allowUserInteraction, animations: {
            self.disconnectButton.layer.opacity = 1.0
        })
    }

    func startTimer() {
        timer = Timer.scheduledTimer(timeInterval: TimeInterval(TIMER_INTERVAL), target: self, selector: #selector(TwilioVideoViewController.hideDisconnectButton), userInfo: nil, repeats: false)
    }

    func resetTimer() {
        stopTimer()
        startTimer()
    }

    func stopTimer() {
        timer?.invalidate()
        timer = nil
    }

// MARK: - Public

    @IBAction func disconnectButtonPressed(_ sender: Any) {
        stopTimer()
        room?.disconnect()
        delegate?.dismiss()
    }

// MARK: - Private
    func doConnect(_ room: String) {
        if (accessToken == "TWILIO_ACCESS_TOKEN") {
            logMessage("Please provide a valid token to connect to a room");
            return
        }

        let connectOptions = TVIConnectOptions(token: accessToken, block: { builder in

                // The name of the Room where the Client will attempt to connect to. Please note that if you pass an empty
                // Room `name`, the Client will create one for you. You can get the name or sid from any connected Room.
                builder.roomName = room
            })

        // Connect to the Room using the options we provided.
        self.room = TwilioVideo.connect(with: connectOptions, delegate: self)

        logMessage("Attempting to connect to room \(room)")
    }

    func setupRemoteView() {
        // Creating `TVIVideoView` programmatically
        let remoteView = TVIVideoView()

        // `TVIVideoView` supports UIViewContentModeScaleToFill, UIViewContentModeScaleAspectFill and UIViewContentModeScaleAspectFit
        // UIViewContentModeScaleAspectFit is the default mode when you create `TVIVideoView` programmatically.
        self.remoteView?.contentMode = UIView.ContentMode.scaleAspectFit

        view.insertSubview(remoteView, at: 0)
        self.remoteView = remoteView

        var centerX: NSLayoutConstraint? = nil
        if let remoteView = self.remoteView {
            centerX = NSLayoutConstraint(item: remoteView, attribute: .centerX, relatedBy: .equal, toItem: view, attribute: .centerX, multiplier: 1, constant: 0)
        }
        if let centerX = centerX {
            view.addConstraint(centerX)
        }
        var centerY: NSLayoutConstraint? = nil
        if let remoteView = self.remoteView {
            centerY = NSLayoutConstraint(item: remoteView, attribute: .centerY, relatedBy: .equal, toItem: view, attribute: .centerY, multiplier: 1, constant: 0)
        }
        if let centerY = centerY {
            view.addConstraint(centerY)
        }
        var width: NSLayoutConstraint? = nil
        if let remoteView = self.remoteView {
            width = NSLayoutConstraint(item: remoteView, attribute: .width, relatedBy: .equal, toItem: view, attribute: .width, multiplier: 1, constant: 0)
        }
        if let width = width {
            view.addConstraint(width)
        }
        var height: NSLayoutConstraint? = nil
        if let remoteView = self.remoteView {
            height = NSLayoutConstraint(item: remoteView, attribute: .height, relatedBy: .equal, toItem: view, attribute: .height, multiplier: 1, constant: 0)
        }
        if let height = height {
            view.addConstraint(height)
        }
    }

    // Reset the client ui status
    func showRoomUI(_ inRoom: Bool) {
        // self.micButton.hidden = !inRoom;
        // self.disconnectButton.hidden = !inRoom;
        UIApplication.shared.isIdleTimerDisabled = inRoom
    }

    func cleanupRemoteParticipant() {
        if let viewedParticipant = viewedParticipant {
            if viewedParticipant.videoTracks.count > 0, let remoteView = remoteView {
                viewedParticipant.videoTracks[0].videoTrack?.removeRenderer(remoteView)
                remoteView.removeFromSuperview()
            }
            self.viewedParticipant = nil
            delegate?.onDisconnected(room?.localParticipant?.identity, participantSid: room?.localParticipant?.sid)
        }
    }

    func logMessage(_ msg: String?) {
        print("\(msg ?? "")")
    }

// MARK: - UITextFieldDelegate

// MARK: - TVIRoomDelegate
    func didConnect(to room: TVIRoom) {
        // At the moment, this example only supports rendering one Participant at a time.

        logMessage("Connected to room \(room.name) as \(room.localParticipant?.identity ?? "??") ... waiting on participant to join")
        messageLabel.text = remoteParticipantName
        viewedParticipant = nil
        for participant in room.remoteParticipants {
            participant.delegate = self
        }
        TwilioMonitor.set(room: room)

        delegate?.onConnected(room.localParticipant?.identity, participantSid: room.localParticipant?.sid)
    }

    func room(_ room: TVIRoom, didDisconnectWithError error: Error?) {
        cleanupRemoteParticipant()
        self.room = nil

        showRoomUI(false)
        delegate?.dismiss()
    }

    func room(_ room: TVIRoom, didFailToConnectWithError error: Error) {
        logMessage("Failed to connect to room, error = \(error)")

        self.room = nil

        showRoomUI(false)
        delegate?.dismiss()
    }

    func room(_ room: TVIRoom, participantDidConnect participant: TVIRemoteParticipant) {
        participant.delegate = self
        logMessage("Room \(room.name) participant \(participant.identity) connected")
    }

    func room(_ room: TVIRoom, participantDidDisconnect participant: TVIRemoteParticipant) {
        logMessage("Room \(room.name) participant \(participant.identity) disconnected")
        if viewedParticipant == participant {
            logMessage("Participant disconnected")
            cleanupRemoteParticipant()
            delegate?.dismiss()
        }
    }

// MARK: - TVIRemoteParticipantDelegate
    func subscribed(to videoTrack: TVIRemoteVideoTrack, publication: TVIRemoteVideoTrackPublication, for participant: TVIRemoteParticipant) {
        logMessage("Participant \(participant.identity) subscribed to video track.")

        if viewedParticipant != participant {
            cleanupRemoteParticipant()
            viewedParticipant = participant
            setupRemoteView()
            if let remoteView = remoteView {
                videoTrack.addRenderer(remoteView)
            }
        }
    }

    func unsubscribed(from videoTrack: TVIRemoteVideoTrack, publication: TVIRemoteVideoTrackPublication, for participant: TVIRemoteParticipant) {
        logMessage("Participant \(participant.identity) unsubscribed from video track.")

        if viewedParticipant == participant {
            if let remoteView = remoteView {
                videoTrack.removeRenderer(remoteView)
            }
            remoteView?.removeFromSuperview()
            cleanupRemoteParticipant()
            delegate?.dismiss()
            // TODO: This will kick us out....some ideas:
            //  1. Search for another participant with a video track (requires saving all participants or tracking in addedVideoTrack)
        }
    }

// MARK: - TVIVideoViewDelegate
    func videoView(_ view: TVIVideoView, videoDimensionsDidChange dimensions: CMVideoDimensions) {
        logMessage("Dimensions changed to: \(dimensions.width) x \(dimensions.height)")
        self.view.setNeedsLayout()
    }
}

class PlatformUtils: NSObject {
    class func isSimulator() -> Bool {
#if TARGET_IPHONE_SIMULATOR
        return true
#endif
        return false
    }
}
